### PR TITLE
Fix #6938: Save trackable status in log activity

### LINKED ISSUE #6938: Travel Bug status reset with screen rotation during cache log
##### Detailed steps causing the problem:
- Open a cache listing in cgeo, enter the menu to add a log to this cache
- (you have to have TBs in your inventory on GC.com) Change the visited/dropped status of any TB you have
- rotate the screen 

##### Actual behavior after performing these steps:

Depending of the setting at _Settings_ -> _Logging_ -> _Set trackables to "Visited"_, Travel Bug(s) status is always reset to "Do Nothing" (if setting is disabled) or "Visited" (if setting is enabled) each time the screen rotates.

##### Expected behavior after performing these steps:

TB status should not be reset if user changed a given value. Everything else (log type such as Note, Found it, Did not find…, log text, date) is kept between screen rotations.

##### Version of c:geo used:

`2018.02.05`

##### Is the problem reproducible for you?
Yes

##### System information:
```
--- System information ---
Device: D5803 (D5803, Sony)
Android version: 5.0.2
Android build: 23.1.A.1.28
c:geo version: 2018.02.05
Google Play services: disabled - 11.9.75 (236-182402865)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide own/found: false
Map strategy: auto
HW acceleration: enabled (default state)
System language: fr_FR
System date format: dd/MM/y
Debug mode active: no
System internal c:geo dir: /data/data/cgeo.geocaching (680,9 MB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (680,9 MB free) external non-removable
Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (680,9 MB free) external non-removable
Database: /data/data/cgeo.geocaching/databases/data (9,4 MB) on system internal storage
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
   geocaching.com: Logged in (Connecté) / PREMIUM
Geocaching.com date format: d/MM/yyyy
Installed c:geo plugins:  none
--- End of system information ---
```

##### Other comments and remarks:

#6869 is a very similar issue but for the <kbd>[WP Calc](https://github.com/cgeo/cgeo/labels/WP%20Calc)</kbd>

### PR DESCRIPTION
**Pull request by bekuno**
implement solution for Issue #6938 "Travel Bug status reset with screen rotation during cache log"

 - add list of the TB IDs and their state in onSaveInstanceState
 - load this list in onCreate
 - set the state to saved state or to possible set setting AutoVisit

### CODE REVIEW

**Comment by bekuno** [comment]
Question:
The method initializeTrackablesAction is used in row 166 and 432 and called both when initialize the class (call onCreate)

I do not see the need to call it in row 432.
Can someone  check this?

**Comment by Lineflyer** [comment]
ok to test

**Comment by bekuno** [comment]
There is no test class in "LogCacheActivityTest.java".
Can someone guide/help me to create/configure one?

**Comment by Lineflyer** [comment]
I personally think, that this kind of functional change does not necessarily requires a test. However someon with more code knowledge should review and decide.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 1]
```diff
@@ -215,9 +217,20 @@ private void showErrorLoadingAdditionalData() {
...
+            if (trackableState == null) { // initial call
+                if (Settings.isTrackableAutoVisit()) {
+                    trackable.action = LogTypeTrackable.VISITED;
+                }
+            } else { // refresh view
+                final Integer tbStateId = trackableState.getInt(trackable.trackCode);
+                if (tbStateId > 0) { // found in saved list
+                    trackable.action = LogTypeTrackable.getById(tbStateId);
+                } else { // not found
+                    if (Settings.isTrackableAutoVisit()) {
```
Comment (UniQP) [inline main/src/cgeo/geocaching/log/LogCacheActivity.java]: I don't like the duplicated condition. Alternative loop:
```
        for (final TrackableLog trackable : trackables) {
            if (trackableState != null) { // refresh view
                final int tbStateId = trackableState.getInt(trackable.trackCode);
                if (tbStateId > 0) { // found in saved list
                    trackable.action = LogTypeTrackable.getById(tbStateId);
                    continue;
                }
            }
            if (Settings.isTrackableAutoVisit()) {
                trackable.action = LogTypeTrackable.VISITED;
            }
        }
```

File: main/src/cgeo/geocaching/log/LogCacheActivity.java [hunk 2]
```diff
@@ -215,9 +217,20 @@ private void showErrorLoadingAdditionalData() {
...
-        if (Settings.isTrackableAutoVisit()) {
-            for (final TrackableLog trackable : trackables) {
-                trackable.action = LogTypeTrackable.VISITED;
+        for (final TrackableLog trackable : trackables) {
+            if (trackableState == null) { // initial call
+                if (Settings.isTrackableAutoVisit()) {
+                    trackable.action = LogTypeTrackable.VISITED;
+                }
+            } else { // refresh view
+                final Integer tbStateId = trackableState.getInt(trackable.trackCode);
```
[the remark this answers is not in the document]
Comment (bekuno) [reply main/src/cgeo/geocaching/log/LogCacheActivity.java]: Changed.

File: main/src/cgeo/geocaching/log/LogCacheActivity.java (hunk 1 above)
Comment (bekuno) [reply main/src/cgeo/geocaching/log/LogCacheActivity.java]: I don't like the duplicated condition too. 
Changed.

**Comment by UniQP** [comment]
@bekuno Github scans the commit message for keywords to automatically close the corresponding issue: https://help.github.com/articles/closing-issues-using-keywords/

Can you adapt the commit message accordingly? It would also be nice if the commit title summarizes how you solve the issue. Suggestion:
```
Fix #6938: Save trackable status in log activity

…
```

**Comment by bekuno** [comment]
I do not have a gcvote account. And currently (or for ever) the website is down. 
Therefore the implementation is only for TBs, not for the gcvote status.
The issue #6938 is so only partially fixed.

Is it possible to change the commit message later?

**Comment by bekuno** [comment]
After code review - it seems the gcvote rating is already implemented.
(in onSaveInstanceState: outState.putDouble(SAVED_STATE_RATING, rating);)
Or am I wrong?

**Comment by UniQP** [comment]
#6938 did not mention GCVote.

You can change the commit message via interactive rebasing (see https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).

That's easier without the merge commit, so let's rebase on master first
```
git rebase master TB_log_save_state
```
(You can omit the last parameter if your current branch already is `TB_log_save_state`.)

Now you can change all commits of your branch by
```
git rebase -i master
```

You now get an overview of your commits relative to the `master` branch:
```
pick 2638d5b1f implement solution for Issue #6938 "Travel Bug status reset with screen rotation during cache log"
pick 1bc8b1745 resolve issue from codacy
pick 17a678f9a code optimization
```

We want to change the commit message of the first commit, so we change `pick` into `reword` in the first line. Since the other two commits are cleanup commits (which are great to see the incremental changes during the review of the PR but not so nice now after the review is finished), you can combine them with the previous commit by changing `pick` to `fixup`:
```
reword 2638d5b1f implement solution for Issue #6938 "Travel Bug status reset with screen rotation during cache log"
fixup 1bc8b1745 resolve issue from codacy
fixup 17a678f9a code optimization
```
After you've saved the changes and closed the editor, you get a new editor window to change the commit message.

When you try to push your changes, you will notice that `git push` will not work. That's because we didn't only add new commits to the branch. You can push the branch via
```
git push --force-with-lease
```

**Comment by bekuno** [comment]
#4697 is the one with GCVote. Sorry.

Thanks for the detaild steps and extra information to change the commit messages.
Now it is fine (I hope so)

**Comment by UniQP** [comment]
@bekuno Thank you for your contribution

**Pull request merged**